src/include/miopen/adaptivemaxpool/problem_description.hpp

```diff
@@ -0,0 +1,246 @@
+                          const TensorDescriptor& outputDesc_,
+                          const TensorDescriptor& indicesDesc_)
+        : inputDesc(inputDesc_), outputDesc(outputDesc_), indicesDesc(indicesDesc_)
+    {
+        IsValidLength();
+        IsValidDims();
+        IsSameType();
+    }
+
+    auto GetInputDesc() const { return inputDesc; }
+    auto GetOutputDesc() const { return outputDesc; }
+    auto GetNtotal() const { return outputDesc.GetElementSize(); }
+
+    bool IsValidLength() const
+    {
+        auto input_dims = inputDesc.GetLengths().size();
+        if(outputDesc.GetLengths()[0] != inputDesc.GetLengths()[0] ||
+           outputDesc.GetLengths()[1] != inputDesc.GetLengths()[1] ||
+           outputDesc.GetLengths().size() != input_dims)
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "AdaptiveMaxPool: Input and output tensor sizes do not match.");
+        }
+
+        if(input_dims == 3)
+        {
+            if(outputDesc.GetLengths()[2] > inputDesc.GetLengths()[2])
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "AdaptiveMaxPool: Input tensor sizes are too small compare to output "
+                             "tensor sizes.");
+            }
+        }
+        else if(input_dims == 4)
+        {
+            if(outputDesc.GetLengths()[2] > inputDesc.GetLengths()[2] ||
+               outputDesc.GetLengths()[3] > inputDesc.GetLengths()[3])
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "AdaptiveMaxPool: Input tensor sizes are too small compare to output "
+                             "tensor sizes.");
+            }
+        }
+        else if(input_dims == 5)
+        {
+            if(outputDesc.GetLengths()[2] > inputDesc.GetLengths()[2] ||
+               outputDesc.GetLengths()[3] > inputDesc.GetLengths()[3] ||
+               outputDesc.GetLengths()[4] > inputDesc.GetLengths()[4])
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "AdaptiveMaxPool: Input tensor sizes are too small compare to output "
+                             "tensor sizes.");
+            }
+        }
+        if(indicesDesc.GetElementSize() != 1)
+        {
+            if(outputDesc.GetLengths() != indicesDesc.GetLengths())
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "AdaptiveMaxPool: Indices and output tensor sizes do not match.");
+            }
+        }
+
+        return true;
+    }
+
+    bool IsValidDims() const
+    {
+        if(inputDesc.GetLengths().size() > 5 || inputDesc.GetLengths().size() < 3)
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "AdaptiveMaxPool: Only 3D, 4D and 5D tensors are supported.");
+        }
+
+        return true;
+    }
+
+    bool IsAllContiguous() const { return inputDesc.IsContiguous() && outputDesc.IsContiguous(); }
```

Comment:
how about `indicesDesc`? If only `inputDesc` and `outputDesc` are checked then the function name is somewhat misleading. same as backward

Reply:
resolved